### PR TITLE
cross platform typed array proposal

### PR DESCRIPTION
**Pull request by ncannasse**
This is a proposal to implement typed arrays in a cross platform manner, following the JS way of doing.
I'm trying to take proper attention at correctly integrating it together the bytes "views" with the static buffer which is haxe.io.Bytes.

Please note that in order to support this I had to change the haxe.io.BytesData in JS from Uint8Array to ArrayBuffer, which match better our definition. This will most likely require additional compat support for IE8.

### CODE REVIEW

**Comment by ncannasse** [comment]
Ping @waneck @aduros @hughsando

**Comment by ncannasse** [comment]
Please note that I'm not exactly following the JS spec.
 - our typed arrays does not extend ArrayBufferView, you need to access it through the `view` field
 - I added `sub(pos,length)` instead of `subarray(begin,end)` which I find a lot more easy to work with. We can also add subarray as well for compatibility purpose
 - I didn't add all the ES6 API, only what's currently supported in Chrome

Some other notes:
 - This requires an efficient CPP implementation
 - I'm not sure yet how we should specify bounds checks, in particular when writing outside of the view

**Comment by waneck** [comment]
I'm a little confused with this. First of all, I don't think this API will ever be as efficient as it can be on other platforms other than JS. This seems to follow too closely the JS specs - which means that we'll probably need to wrap the bytes data into an object (like you did with ArrayBuffer), which is quite an overhead if you're setting an information byte-by-byte.

I've toyed with an implementation some time ago - it's available at https://github.com/waneck/taurine-hx/blob/sandbox-old/src/taurine/mem/RawMem.hx . You can probably use the tests available at https://github.com/waneck/taurine-hx/blob/sandbox-old/test/taurine/tests/mem/RawMemTests.hx for a more complete set of tests, which by the way we can reuse for this implementation.

I think the fundamental difference is that my implementation only provides a way to access a set of bytes as straight-forward as possible. Note that it doesn't provide a `haxe.io.Bytes` interface outright, because on some platforms (for example, Java), the `haxe.io.BytesData` type is not the same as the fastest way to access any data type from memory (other than a byte).
IMHO, with a `RawMem` access primitive that is guaranteed to be very fast, we can solve the problem that "array views" want to solve - which is of a comfortable way to deal with structs - without needing to wrap anything. Actually I've coded a macro that provides compile-time access to a defined memory layout. If this is interesting, we can maybe add this as a full feature in Haxe - not depending on macros. You can see some of the tests for this at https://github.com/waneck/indian/blob/master/tests/src/indian/test/StructTests.hx#L40 . This was more geared towards native libraries interop, so it's geared towards the differences of each type size in each platform (e.g. the alignment differences on each platform). Also the underlying type - which isn't the same as the `RawMem` I mentioned earlier - is geared towards accessing an unmanaged (not GC'ed) memory chunk.

But you can get the idea that all of this can be achieved for GC'ed memory access only with a very fast `RawMem` primitive: The offsets can be calculated at compile-time, so there's no need to create views or wrap the buffer in any way.

Some other questions:
* Why Uint8Array and Float32Array only?
* I saw there's a byte offset there but it doesn't seem to be set anywhere.

**Comment by waneck** [comment]
One interesting but unrelated code that may be useful to the Haxe std lib, and was mentioned on the latest wwx is the work I did with indian on encoding: https://github.com/waneck/indian/tree/master/src/indian/types/encoding . There were some tests at https://github.com/waneck/indian/blob/master/tests/src/indian/test/UnicodeTests.hx and they are all passing on the indian targets - C#, Java, C++ and Neko. It even supports UTF-16 surrogate pairs.

**Comment by ncannasse** [comment]
@waneck I'm not sure I understand your concerns regarding performances. Is it because of the additional offset required by the "view" paradigm ? I don't think on modern CPUs using a runtime offset is slower than using a compiletime generated constant, and it is definitely useful to be able to make different views on the same underlying data for CPU cache speedup

**Comment by ncannasse** [comment]
I only added Uint8Array and Float32Array atm until we validate the structure. Implementing the others should be quite straightforward after we agree on these two :)

**Comment by ncannasse** [comment]
@waneck another question : how you RawMem differs from haxe.io.Bytes ? My idea is to have haxe.io.Bytes stand for the default "raw buffer" API, then have more high level APIs as views on this raw buffer

**Comment by Simn** [comment]
What I'm confused about is that everything is still based on haxe.io.Bytes. I thought the whole idea was to get rid of its data wrapping...

**Comment by ncannasse** [comment]
@Simn any idea about the strange python failing test ?

**Comment by ncannasse** [comment]
@Simn we're only using haxe.io.Bytes atm for "fallback" default implementation. I have not yet added Neko/C++ specific implementations, but they will use the lowest unwrapped data. From an user POV however, haxe.io.Bytes will still be the underlying raw buffer.

**Comment by Simn** [comment]
No idea, maybe @nadako or @frabbit can help.

**Comment by underscorediscovery** [comment]
**I am excited for typed arrays in the std api!** I was making a "standalone" of the es5/es6 stable implementation for other haxe users and frameworks (snow,lime,etc) to share, so to remove that need would be great. 

However, I wanted to retain the usage patterns for the end user from the es6 spec, so I have opted to maintain my existing implementation as a library, which is (and would remain) compatible with haxe 3.1.3 as a bonus. When the final implementation of std buffers arrives, it can switch to leverage that accordingly.

Nicolas and I briefly discussed working it into std, but he opted for a more direct custom api (not a bad thing) - I felt it pertinent to mention it in this thread all the same (as it includes some prelim discussion in the issues list).

The link is here, for the curious. https://github.com/underscorediscovery/hxtypedarray

**Comment by Simn** [comment]
Are you guys serious? I was glad to see that @underscorediscovery was working on this while in communication with Nicolas, but it looks like that didn't turn out so well. I really don't want to have two very similar but incompatible implementations of this. We're already having enough fragmentation as it is.

Could we please work this out? What are the fundamental problems here?

**Comment by underscorediscovery** [comment]
Not sure if there is anything to 'work out'...

My code was used all along. Originally ported to lime back in 2013, refactored and wrote unit tests in snow, [which is still current](https://github.com/underscorediscovery/snow/tree/master/snow/platform/native/utils) and in use by all luxe/snow users. [July 2014](https://github.com/underscorediscovery/snow/issues/26) I did some work, but needed to root out the ByteArray code. I decided to wait for the new 3.2 Magic for hxcpp.

Lately though, I redid them as abstracts, Joshua gave me the idea. I did this to use the cpp memory copy and unifying the IO layers. [This was done here](https://github.com/underscorediscovery/snow/tree/fasttypedarrays/snow/platform/native/utils) and saw huge performance over the old.

After that is when Nicolas and I spoke, he gave me some insight/feedback and suggested it would ideally be over haxe.io.Bytes as a start, and I could PR for discussion. I refactored to do that. He decided instead to go with a new api - I trust it will be good in the end, as I trust haxe team overall it's why I am here. 

In any case, this PR doesn't change my course. I have my code in a better form, I need it now, I need it to be what it is. That isn't going to change, and when 3.2 comes along, course can be adjusted to make the best use of the available api.

**Comment by hughsando** [comment]
I agree with @waneck that this is not an optimal solution for fast byte access.  But then it is not meant to be.  You need to be able to create multiple ArrayBufferViews at different offsets pointing to a single data source.  This necessitates a class that has a data pointer and an offset at the very least.  So that is an extra alloc per object, and an extra index per access.

If you make an abstract view over BytesData, then I think you will have something like maximum seed on hxcpp.  Say ```abstract FloatBytes(BytesData) implements ArrayAccess<Float>``` could be made to work at 100% speed.  And another abstract that lets you specify the type/poistion ("getInt32At(pos)") for mixed/struct access is going to allow the fasted possible access. All endian would be specified as "Native". If you allow the actual type to be target dependent, and do not require "toBytes" or anything, then I think Cauê will be able to generate optimal code.  But I guess this is not what we are talking about are we.  (although I would like a standard solution for this too).


So there will be 3 objects for the ArrayBufferViewData/Bytes/BytesData in hxcpp - I think you can get away with 2 If we could make an "ArrayBufferData" to be build on "BytesData" rather than "Bytes".  I also think the "ArrayBufferView" should be exposed as a first-class entity.

From my point of view, the only use for this class is for write js-compatible code - ie, WebGL.  And this needs ArrayBufferView, so it should be there, and you should add as many matching functions as possible.

My preferred implementation would go like this:
 1. define a "BytesAccess"  or "ArrayBufferData" abstract with target dependent data that has "getFloat32At(pos)/getInt32At(pos) and maybe "sub/blit".  Do not encumber with with any "toBytes" functions that define the implementation.
2. Build ArrayBufferView that "hasA" ArrayBufferData and adds the js functions, and maintains an "offset"
3. Float32Array could "extend" rather than abstract the view - this would give you reflective access to "set" - but not too fussed about this.
4. Expose the ArrayBufferData property for direct access.

So, lets say I want to build a "struct" to pass to WebGL : ```{x:Float, y:Float, index:Int}```.
I would probably store the "ArrayBufferData" in my class.  On hxcpp, this would be an ```Array<unsigned char>``` - pretty much the minimal allocation you can do - a single object.
I would then do something like:
```
   var location = 0;
   for(obj  in objects)
   {
        data.setFloat32(location, obj.x); location+=4;
        data.setFloat32(location, obj.x);  location+=4;
        data.setInt32(location,obj.index); location+=4;
   }

   someWebGlFunction( ArrayBufferView.ofData(data) ); // Or cache ArrayBufferView, and use "data=view.getData()"
```
This code would be 100% optimal on cpp (java,cs too I think), and I think the API is very nice and simple.

**Comment by ncannasse** [comment]
@Hugh if we follow what you're telling, then it means turning haxe.io.Bytes into an abstract should already do most of this, is that alright ? (or do you propose BytesData not to have for instance any length information ?)

I'm trying the find the best solution that brings both efficient speed in all the platforms and a nice API for the end user.

Regarding @underscorediscovery comments, if I understood correctly he needs a solution which works right now and is compatible with Haxe 3.1. I don't see any issue that he uses his own solution as long as he agree to switch to the std one once 3.2 is released.

**Comment by jgranick** [comment]
My primary issue with implementing the typed array spec were the constructors -- you can make a `Float32Array` out of a standard array, with an `Int` length, or out of another `ArrayBuffer`. I felt that using abstracts (or macros? or generic?) we should be able to optimize these constructor paths, but had to use Reflection at runtime instead :cry: 

I don't think that the "view" (`Float32Array`, `Int8Array`, etc) _need_ to be abstract instead of classes, but it had a big bonus for treating these as a read/write interface.

`ArrayBuffer` is a blob of data. Every "view" is a way to read and write the data. I don't think it matters what kind of blob `ArrayBuffer` is made of (so long as you standard ways of constructing it) and the views, IMHO, should be as minimal as possible, as it's really syntax around the I/O for this, similar to `ByteArray` or `Bytes` but divided into multiple interfaces.

**Comment by jgranick** [comment]
BTW, the `getFloat32`, `setFloat32` type interface that @waneck mentions would be a `DataView`, the `ArrayBufferView` implementations could use this internally (or not) if it cleans up the code

**Comment by hughsando** [comment]
I think haxe.io.Bytes as an abstract would remove the "extra object"
problem.  For hxcpp, the BytesData (```Array<unsigned char>```) has a
length already accessible. Not sure about the other targets.  The fast
access is already implemented for BytesData:

https://github.com/HaxeFoundation/hxcpp/blob/master/include/hx/StdLibs.h#L213

These could be exposed as abstracts on Bytes, and you would have a 100%
speed option.  For what I understand, this is not the cast on Java, because
you would need to use something different for BytesData - but maybe this is
possible.

Having identified the "fastest" way to access raw bytes, we could then look
at the "most convenient", which I guess involves presenting the bytes in
ways usable by external Apis - BytesInput/Output, ArrayBufferView,
IDataInput,IDataOutput, ByteArray, Float32Array etc.  nme did this via:
```class ByteArray extends Bytes implements ArrayAccess<Int> implements
IDataInput implements IMemoryRange implements IDataOutput```  And used
wrapper objects for the views.
So this "convenience" object owns (by inheritance) the BytesData and you
have a pretty convenient object, with only 1 "extra object".  I am not
saying this is the greatest way of doing things, but giving a hint as to
what a convenient object(s) should be able to do.

I guess if we work with something like "various convenience objects can
share references to the single fast-memory-access abstract" then the
developer can work with 100% speed, and interop to any of the extenal APIs
via lightweight wrappers.  It then does not really matter about the speed
of these wrappers, since the developer an always use the "well-defined fast
way".




On Tue, Feb 3, 2015 at 3:07 AM, Nicolas Cannasse <notifications@github.com>
wrote:

> @Hugh <https://github.com/Hugh> if we follow what you're telling, then it
> means turning haxe.io.Bytes into an abstract should already do most of
> this, is that alright ? (or do you propose BytesData not to have for
> instance any length information ?)
>
> I'm trying the find the best solution that brings both efficient speed in
> all the platforms and a nice API for the end user.
>
> Regarding @underscorediscovery <https://github.com/underscorediscovery>
> comments, if I understood correctly he needs a solution which works right
> now and is compatible with Haxe 3.1. I don't see any issue that he uses his
> own solution as long as he agree to switch to the std one once 3.2 is
> released.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/haxe/pull/3827#issuecomment-72516906>.
>

**Comment by underscorediscovery** [comment]
I would note that it is not _purely_ "convenience", abusing the cache will always be slower, as nicolas already noted, the views fit the need to share data in a cache friendly way. 

It's the same reason you have chunks pre-allocated AOT, but instead of fiddling with byte values for days it's a user friendly API for even the not so technical. It also plainly separates "length" from being ambiguous, which has high value in clarity. length = elements, byteLength = well....

**Comment by Justinfront** [comment]
On 03/02/2015 00:13, Hugh Sanderson wrote:
> or what I understand, this is not the cast on Java, because
> you would need to use something different for BytesData
Azrafe7 implemented a Pixel class that works on js, java, nme targets 
for portrace.  Being that Haxe is so games focused I presume real use 
case of pixel manipulation should be considered cross platform in 
relation to Bytes processing, so maybe the link is useful in 
consideration, ie what would this look like in respect to solutions 
proposed, just don't want you guys to get too abstract without 
consideration of us abusers of your work :)

https://github.com/Justinfront/hxDaedalus/blob/master/src/hxDaedalus/graphics/Pixels.hx

**Comment by hughsando** [comment]
Yes, @underscorediscovery, I'm not suggesting we all program in assembler, but if all these functions are built on top of the same fastest-memory-access, we can have a copy-free Api that still allows 100% speed.  For example, the ArrayBufferView necessitates an "offset".  Fast code may wish to combine this offset with its own offset to avoid a member-lookup-addition per access.  I'm not saying that everyone should do this, but being able to do this when profiling suggests you should, is part of a good Api.

**Comment by underscorediscovery** [comment]
Yea @hughsando, I may have unintentionally implied something, I was mainly thinking aloud that the ES spec has been good for js _users_. Agreed on the rest, though.

**Comment by ncannasse** [comment]
Ok, so let's decide what we need to do here: if we turn haxe.io.Bytes into and abstract, would that be enough for now for you @hughsando and @waneck ?

Regarding Java, if BytesData is not the fastest way to implement IO access, can't we change that so it becomes so ?

Another question : hugh mentioned that we should introduce a proper common type ArrayView. Any arguments in favor of this solution ? I feel that doing pseudo inheritance with abstracts is not really the proper way to do so and would somehow encourage users to unsafe-cast to the real type (which is error prone) but I would like to hear arguments in favor of the "subtyping" paradigm

**Comment by hughsando** [comment]
Yes, I think Bytes as an abstract should allow full speed.  But may also break code.  We could simply add a new abstract - not in "io" - "haxe.Memory" or something like that.
I was thinking that neko could use a native type that binds the "length" with the string data and reallocation once we decide the haxe api - maybe it already has this?
One more thing, if we are changing the Api, is to avoid the function called "readByte" or probably anything involving "Byte" because flash (wrongly, IMHO) defines that as signed, and this is confusing and causes bugs when switching between Apis.  uint8/int8 are not ambiguous.

**Comment by ncannasse** [comment]
@hughsando turning Bytes into an abstract should not break much user code, I'm in favor of keeping things simple here (which means avoiding introducing several types which does the same thing). What kind of break are you expecting here ?

I agree with the readByte comment, but this will definitly cause a lot of breakage if we deprecate haxe.ioInput/Output read/writeByte, so let's only introduce readUInt8 for now and we can change that later (haxe.io.Bytes is using get/set). I don't think that we should change haxe.io.Bytes name here.

For neko : yes the string basic type already has its length and can be manipulated as binary. However it can't be resized and we must not assume that haxe.io.Bytes can be reallocated/resized. So I propose we keep things this way:

- haxe.io.Bytes is a abstract API for manipulating BytesData
- haxe.io.BytesData is the single static memory chunk (fixed size)
- haxe.io.UInt8Array (and friends) are views on a part of this memory
- haxe.io.DataView as well (similar API to Bytes but is a view and explicit endianness)
- haxe.io.BytesBuffer allows writing incrementally in an efficient manner, does take care of grow/resize in an efficient manner and when done produce an haxe.io.Bytes 
- haxe.io.Input/Output are slower but provides an higher level of abstraction (can works on bytes, socket, files, etc.)

Would that work for everybody ?

**Comment by frabbit** [comment]
is the python problem still valid? I cannot reproduce it. Here is the output for the command `git log | head; python3.2 -V; python3.2 unit.py`:

```

commit 06068a529896ad40066cb3a5796fcb8c834925d4
Author: Nicolas Cannasse <ncannasse@gmail.com>
Date:   Mon Feb 2 00:10:19 2015 +0100

    more work on ie8-9 emulation

commit dcf67c32575e81b8757fd452372e688432f019d8
Author: Nicolas Cannasse <ncannasse@gmail.com>
Date:   Sun Feb 1 22:47:06 2015 +0100

Python 3.2.3
Test.hx:244: Generated at: 2015-02-04 19:11:51
Test.hx:246: START
Test.hx:187: DONE [5474 tests]
Test.hx:353: SUCCESS: true
```

**Comment by Simn** [comment]
Looks like it was fixed somewhere along the way. Thanks for double-checking!

**Comment by waneck** [comment]
Hi! Sorry for taking some time to answer!

I think the solution you're proposing is good. Making `haxe.io.Bytes` an abstract does solve some issues for me. However I agree with Hugh that it might break a lot of code - for example, our own Serializer checks if `Std.is(value,Bytes)`, which won't work if we're dealing with abstracts. Maybe the "Memory" abstract would be the way to go.

Making it an abstract would be nice to me - so it doesn't need wrapping and it can have a different underlying type than BytesData for Java, which is what I'd need to make this work as efficiently as possible. There are some problems with changing Java's BytesData type - I can go more on details later if you'd like.

Overall, I mostly agree with Hugh on his points. The structure example that he was proposing was kind of what I meant to say with that `RawMem` and it's how I'm dealing with structs on indian. I really think that provide a very lean implementation on top of which we can create more convenient abstractions (like the typed arrays themselves - which are great to have in the std lib! -, or a macro-based struct) would be the best option. I've been working for some time now on getting structs to work nicely and efficiently with Haxe, and having a "Memory" abstract which is very fast will be the building block to something on this direction. Perhaps then in the future (if there is interest) these kinds of structs could be even implemented on a language level instead of a macro-based solution. But this can be discussed in the future :)

One issue that may be related is about GC and memory ownership, Right now on Neko and Hxcpp there's no way to access bytes that are not owned (created) by the GC. This works fine until we need to e.g. access memory mapped files, or work on some api-specific buffer. It would be nice that since we're reviewing how to deal with bytes, that maybe we think of a way that could solve this as well. It would certainly be great if there was the possibility of accessing a memory mapped file through that same "Memory" interface!

**Comment by hughsando** [comment]
Hi Caue - good to hear your input.
Currently, you can access raw/external data using the cpp.Pointer class.
And you can deliver the pointer via a Dynamic "kind", which works nicely
with cffi.

I have just come up with a "trick" that allows you to easily access
externally allocated class pointers, while also being able to mix with
Dynamic.  And with a similar trick, I can extend this structs that have the
same copy-by-value semantics (Like Int/Float/Bool/String), and also mix
with Dynamic by doing a boxed-copy.  While in haxe classes, or on the
stack, these values will not "alloc" and objects, so these should be useful
for the Int64 implementation too.  Have not worked out the details 100%,
but currently they require the struct to be defined in a c++ header file
(ie external, but you could use some meta-magic to generate the headers),
but a language feature would be super handy too.
Such a class can't be built with a abstract over BytesData because this
would imply an actual BytesData object whereas it would actually be
implemented on stack/heap access.
To be really useful, would need to work out how to mix in arrays of these
objects too.

Could you elaborate a bit on the differences between an optimised BytesData
and optimized MemoryData structures?

Hugh

**Comment by ncannasse** [comment]
@waneck I don't think a lot of people are doing a lot of Dynamic checks such as Std.is(haxe.io.Bytes). Anyway by changing to an abstract this will give a proper compilation error "Cannot use abstract as value" so we only need to give an alternative (well optimized per platform) to check if a Dynamic is a BytesData, and document this change.

I'm not in favor of having two different types for doing the same thing. I don't really see the point of keeping an old "slower" version of "RamMem" in haxe.io.Bytes.

**Comment by nadako** [comment]
By the way, what about #2976?

**Comment by ncannasse** [comment]
@nadako it's been postponed as "long term" feature enhancement. Let's not make this case more complicated than it is already :)

**Comment by delahee** [comment]
methinks this issue cristallises too many energy, may i suggest we bail out
till next version so we can have a fresh haxe soon ?

2015-02-05 15:46 GMT+01:00 Nicolas Cannasse <notifications@github.com>:

> @nadako <https://github.com/nadako> it's been postponed as "long term"
> feature enhancement. Let's not make this case more complicated than it is
> already :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/HaxeFoundation/haxe/pull/3827#issuecomment-73057239>.
>



-- 
David Elahee

**Comment by Simn** [comment]
I'm also a bit nervous about this. I know it looks like it's "almost there", but that could be quite deceptive. On the other hand we'll probably only figure out potential problems if we release this to the public at some point, which is the same problem we're always having with stable/unstable changes.

**Comment by jgranick** [comment]
My two cents:

Stable/fast typed arrays is a cornerstone to Lime 2 as well as snow. Any Haxe internal fixes required to support this are better done by this release (if possible) because we'll just have to sit on our hands for another year. I'm not assuming that bug fixes need to be made, but in case there is one, it can block things until the next major release.

@underscorediscovery has a pretty robust set of tests for the typed array implementation, which could be used for this as well. Even if it does not get included officially in this release, any of the core groundwork (making sure there are "fastest" memory management paths for each target, ensuring there are no abstract/generic/low-level bugs that crop up when compiling to C++ or another target) will help third-party implementations. So the "heart" of it is crucial, even if the actual end implementation doesn't get finalized yet

**Comment by delahee** [comment]
Well well last haxe release is so old...any issue taking 40 ping without
consensus can wait. Imvho.
Le 5 févr. 2015 18:41, "Simon Krajewski" <notifications@github.com> a écrit
:

**Comment by ncannasse** [comment]
I agree, we need to move this forward and have it in 3.2. I think we have a consensus on the goals, we're just discussing implementation details, which is not that hard to resolve.

I'm in favor of switching quickly Haxe/GIT to this new version of Bytes so everybody can check which impact it has with 3rd party libs, performances, platform specific details etc. We can still choose to revert these before 3.2 if we think we have not solved everything, but let's not stall the ongoing process.

**Comment by Simn** [comment]
Works for me.

**Comment by delahee** [comment]
Ok for me.
Le 5 févr. 2015 19:02, "Simon Krajewski" <notifications@github.com> a écrit
:

**Comment by underscorediscovery** [comment]
A note: performance impacts might be masked by other changes in hxcpp if someone is testing "fresh"  i.e false flags. I haven't tested other targets too much but I am seeing some change in performance every few git pulls on repo for cpp target, sometimes up, sometimes down. 

As long as the changes are clearly documented, and keep similar functionality (i.e forcing the user suddenly to handle code that std was doing before in Bytes, for all targets, is less nice) - whatever changes land in 3.2 won't affect my end user api, and will positively impact performance in a good way which sounds good.

If people jump onto 3.2 expecting no changes after a year of solid work, they are also probably just missing some notes in the docs.

**Comment by hughsando** [comment]
Maybe you should just call it haxe 4 - given it's been a year and you are
breaking stuff.  You can always do a haxe 5 next year.

**Comment by ncannasse** [comment]
Ok I'll merge this PR and start working on haxe.io.Bytes abstract

**Comment by tondy67** [comment]
from wp
http://en.wikipedia.org/wiki/Asm.js
"Much of this performance gain over normal JavaScript is due to 100% type consistency and virtually no garbage collection (memory is manually managed in a large typed array). This simpler model with no dynamic behavior, no memory allocation or deallocation, just a narrow set of well-defined integer and floating point operations enables much greater performance and potential for optimization."...

File: std/js/_std/haxe/io/Bytes.hx [hunk 1]
```diff
@@ -191,15 +192,18 @@ class Bytes {
...
-		return new Bytes(b.length,b);
+		var hb = untyped b.hxBytes;
+		if( hb != null ) return hb;
+		return new Bytes(b);
 	}
 
 	public inline static function fastGet( b : BytesData, pos : Int ) : Int {
-		return b[pos];
+		// this requires that we have wrapped it with haxe.io.Bytes beforehand
+		return untyped b.bytes[pos];
```
Comment (aduros) [inline std/js/_std/haxe/io/Bytes.hx]: This is throwing a null pointer error for me. Where is the bytes property set? This should probably be b.hxBytes.b[pos].

Adding a hxBytes property onto all ArrayBuffers seems scary...

Comment (ncannasse) [inline std/js/_std/haxe/io/Bytes.hx]: Right. It seems we don't have unit tests for fastGet ;)

**Pull request merged**